speed.py

```diff
@@ -0,0 +1,111 @@
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+
+# from datasets import load_dataset
+
+# dataset = load_dataset("roszcz/internship-midi-data-science", split="train")
+
+
+"""
+SPEED
+"""
+
+
+def get_time_vs_speed(record):
+    df = pd.DataFrame(record["notes"])
+
+    start = df.start.values
+    end = df.end.values
+
+    if (max(end) - min(start)) > 120:
+        step = 60
+        unit = "min"
+    else:
+        step = 1
+        unit = "sec"
+
+    notes_in_time = pd.DataFrame(
+        df.groupby(
+            pd.cut(
+                df["start"], range(int(min(df["start"])), int(max(df["start"])), step)
+            )
+        ).pitch.count()
+    )
+    notes_in_time.index = np.arange(
+        int(min(start)) + 1, len(notes_in_time) + int(min(start)) + 1
+    )
+    notes_in_time.index.name = f"{unit}_number"
+    notes_in_time.rename(columns={"pitch": "number_of_notes_played"}, inplace=True)
+
+    return notes_in_time
+
+
+def plot_chart_time_vs_speed(notes_in_time):
+    unit = notes_in_time.index.name.split("_")[0]
+    step = 1 if unit == "sec" else 60
+
+    f, ax = plt.subplots(1, 1, figsize=[14, 6])
+    ax.scatter(
+        x=notes_in_time.index,
+        y=notes_in_time["number_of_notes_played"].values / step,
+    )
+    ax.set_xlabel(f"Time ({unit})")
+    ax.set_ylabel("Notes played per second")
+    ax.set_title("Time vs Speed")
+    ax.grid(alpha=0.5, color="r", linestyle="--", linewidth=0.25)
+    f.tight_layout()
+
+    return f
+
+
+"""
+NOTES PRESSED AT THE SAME TIME
+"""
+
+
+def get_notes_idxs_pressed_simultaneously(record):
+    df = pd.DataFrame(record["notes"])
+    start = df.start.values
+
+    threshold = 0.05  # can be adjust/changed, eg. to 0.1
+    pressed_simultaneously = []
+    it = 0
+
+    while it < len(df):
+        t_0 = start[it]
+        where_close = (start[it + 1 :] - t_0) < threshold
```

Comment:
```suggestion
        where_close = (start[it + 1 : it + 11] - t_0) < threshold
```

Looking at "everything after" is inefficient - the difference is not huge (1.7s yours, 0.4s mine for 88k notes), but I think it's worth it. 

My suggestion uses only 10 notes, because we're investigating human piano performance :) (Of course it's not impossible for humans to play more than 10 notes at once, so this threshold is debatable 🤔 )